### ISSUE TITLE
Workflow stuck says it's updating workflow but remains in pending state in UI

### ISSUE THREAD

**Issue by snowzach** [argoproj/argo-workflows#6228]
## Summary 

The workflow appears stuck. I tried with 2.11 and 3.0 and the same thing happens. 

## Diagnostics

I keep running into the issue where I run a workflow that previously worked. It used the withParam to branch out into a lot of branches. I recently changed a step where a sub-branch also has a withParam clause so it branches out further. Now it works sometimes but mostly fails. The sub-sub-branches always end up with something stuck and everything else seems to be stuck as well once that happens. The only major change was adding the sub-sub-branch piece. 

It seems like it's trying to work... It keeps saying this in the logs for one of the pods: 
```
time="2021-06-26T03:50:22.134Z" level=info msg="Pod node ztech-reports-2jk78-4032348013 initialized Pending" namespace=argo workflow=ztech-reports-2jk78
time="2021-06-26T03:50:22.236Z" level=info msg="Created pod: ztech-reports-2jk78[2].ztech-report-run-ctrls(10:compound_name:fake,ztech_run_name:IHR20200810,processing_state:15)[6].ztechflow-report-save(0) (ztech-re
ports-2jk78-4032348013)" namespace=argo workflow=ztech-reports-2jk78
time="2021-06-26T03:50:53.645Z" level=info msg="Setting node ztech-reports-2jk78-4032348013 outputs"
time="2021-06-26T03:50:53.645Z" level=info msg="Updating node ztech-reports-2jk78-4032348013 status Pending -> Succeeded"
time="2021-06-26T03:50:54.192Z" level=info msg="SG Outbound nodes of ztech-reports-2jk78-3306362686 are [ztech-reports-2jk78-4032348013]" namespace=argo workflow=ztech-reports-2jk78
time="2021-06-26T03:51:28.522Z" level=info msg="Setting node ztech-reports-2jk78-4032348013 outputs"
time="2021-06-26T03:51:28.522Z" level=info msg="Updating node ztech-reports-2jk78-4032348013 status Pending -> Succeeded"
time="2021-06-26T03:51:28.567Z" level=info msg="SG Outbound nodes of ztech-reports-2jk78-3306362686 are [ztech-reports-2jk78-4032348013]" namespace=argo workflow=ztech-reports-2jk78
time="2021-06-26T03:58:39.464Z" level=info msg="Setting node ztech-reports-2jk78-4032348013 outputs"
time="2021-06-26T03:58:39.464Z" level=info msg="Updating node ztech-reports-2jk78-4032348013 status Pending -> Succeeded"
time="2021-06-26T03:58:39.500Z" level=info msg="SG Outbound nodes of ztech-reports-2jk78-3306362686 are [ztech-reports-2jk78-4032348013]" namespace=argo workflow=ztech-reports-2jk78
time="2021-06-26T03:59:14.180Z" level=info msg="Setting node ztech-reports-2jk78-4032348013 outputs"
time="2021-06-26T03:59:14.180Z" level=info msg="Updating node ztech-reports-2jk78-4032348013 status Pending -> Succeeded"
time="2021-06-26T03:59:14.218Z" level=info msg="SG Outbound nodes of ztech-reports-2jk78-3306362686 are [ztech-reports-2jk78-4032348013]" namespace=argo workflow=ztech-reports-2jk78
```
I think that's the smoking gun, there's nothing really off otherwise about the logs. the workflow-controller doesn't ever seem to realize the pod has compelted.

The kubernetes pod shows completed but it never updates in the UI nor continues with the workflow.

What Kubernetes provider are you using? GKE 1.20

What version of Argo Workflows are you running? Tried in 2.11.12 and 3.0.3 and 3.0.8 with the same issues. 

What executor are you running? argoexec:v2.12.11 (EDIT: using pns for containerd)

Did this work in a previous version? I.e. is it a regression? No, appears to be the same over all of them.

Are you pasting thousands of log lines? That's too much information. 

I will try to come up with something repeatable. The workflow is huge but at a high level its...

workflow A
- step1a - make some json
- step2a - withParam step1a json call workflow B (maybe 30 in parallel)
workflow B
- step1b - make some json
- step2b - withParam 1b json do 3 things

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by snowzach** [comment]
So a little more background here... I am using a Postgres database for archiving... One thing I noticed is that the name of the sub-sub-workflows was absurdly long... So I got to thinking about what might be affected by too long of a workflow step name... It would appear when you use the withParam option the json you pass becomes part of the name of the workflow step. So I went to look into the postgres database and sure enough all of the columns are using `varying(X)` I suspect this might be part of the problem. If I get some time, I will try switching them to `text` and see if that makes a difference.

**Comment by snowzach** [comment]
Okay, so I redesigned the workflow to eliminate the sub-sub-step and make it a single step removing the withParam step. That didn't fix it. I completely removed the archive database. That didn't fix it.

**Comment by snowzach** [comment]
So it thinks it's updating the workflow CRD but it never seems to take... I am wondering if this is some sort of Kubernetes issue. It works most of the time but when it fails, a lot of them fail in the same way. workflow CRD says Pending, kubernetes says pod completed.

**Comment by snowzach** [comment]
Okay... so I went full scorched earth here and completely killed my argo namespace and re-created everything from scratch using the same charts I've been using since I started using argo. (I tested my restoration instructions and they worked. Hooray for the DR plan!) So far it's been humming along perfectly so I'm not sure if that makes me feel better or worse... 

I will leave this open in case a maintainer wants to look or think about this but you all probably have better things to do and bigger fish to fry so feel free to close it. Thanks for an amazing piece of software!

**Comment by alexec** [comment]
CRD=Custom Resource Definition, do you actually mean CR, i.e. a `Workflow`?

**Comment by snowzach** [comment]
> 
> 
> CRD=Custom Resource Definition, do you actually mean CR, i.e. a `Workflow`?

Yes, you're right. Sorry, edited my post.

**Comment by alexec** [comment]
@sarabala1979 this is not the first time I've seen a workflow repeatedly try to update a node to succeeded and the update did not stick. Something to keep an eye on.

**Comment by sarabala1979** [comment]
@snowzach how big is your workflow object which is stuck? you can do `kubectl get workflow <> -o yaml` save in file and check the bytes.  If it is really big which crosses the 1MB k8s limitation, you can try workflows of workflow patterns. the master workflow will submit multiple small workflows.

**Comment by snowzach** [comment]
> 
> 
> @snowzach how big is your workflow object which is stuck? you can do `kubectl get workflow <> -o yaml` save in file and check the bytes. If it is really big which crosses the 1MB k8s limitation, you can try workflows of workflow patterns. the master workflow will submit multiple small workflows.

@sarabala1979 I can measure it if you would think it would be helpful. When I deleted my argo namespace and re-applied my config (`https://raw.githubusercontent.com/argoproj/argo-workflows/v3.0.8/manifests/install.yaml` + my workflow-controller-configmap) without changing anything else, it started working without issue. 

Oh, I realized I am using the pns executor. Not sure if that's relevant.

**Comment by snowzach** [comment]
My config

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: workflow-controller-configmap
  namespace: argo
data:
  containerRuntimeExecutor: pns # Needed to work with containerd

  artifactRepository: |
    archiveLogs: true
    gcs:
      bucket: blahblah01-argo-artifacts
      key: blahblah01_k8s01
      serviceAccountKeySecret:
        name: argo-artifacts
        key: service_account.json

  links: |
    - name: Logs
      scope: workflow
      url: https://console.cloud.google.com/logs/query;query=resource.type%3D"k8s_container"%0Aresource.labels.namespace_name%3D"${metadata.namespace}"%0Alabels."k8s-pod%2Fworkflows_argoproj_io%2Fworkflow"%3D"${metadata.name}"?project=blahblah-1
    - name: Logs
      scope: pod
      url: https://console.cloud.google.com/logs/query;query=resource.type%3D"k8s_container"%0Aresource.labels.namespace_name%3D"${metadata.namespace}"%0Aresource.labels.pod_name%3D"${metadata.name}"?project=blahblah-1
    - name: Example Pod Link
      scope: pod
      url: http://logging-facility?namespace=${metadata.namespace}&podName=${metadata.name}&startedAt=${status.startedAt}&finishedAt=${status.finishedAt}

  workflowDefaults: |
    metadata:
      annotations:
        prometheus.io/scrape: "false"
    spec:
      activeDeadlineSeconds: 28800
      podGC:
        strategy: OnWorkflowSuccess
      ttlStrategy:
        secondsAfterCompletion: 28800
      podDisruptionBudget:
        minAvailable: 9999
      nodeSelector:
        cloud.google.com/gke-nodepool: "genome-generic"
      tolerations:
        - key: "node-pool"
          value: "genome-generic"
          operator: "Equal"
          effect: "NoSchedule"

  metricsConfig: |
    disableLegacy: true
    enabled: true
    path: /metrics
    port: 9090
```

**Comment by snowzach** [comment]
I did check and my refactored workflow ended up being 687k all said and done. I suppose it's plausible it could have hit 1MB before.

**Comment by sarabala1979** [comment]
@snowzach  Can you try to reproduce the same issue with refactored workflow and old workflow? just nail down the root cause.  Each k8s implementation behaves differently when objects reaches the size limits.

**Comment by snowzach** [comment]
Hi @sarabala1979 I tried using both workflow formats (withParam and without) on one of the largest workflows I have and it did not generate the issue. As well, the size of the workflows never exceeded 700k. The issue is definitely gone now. It's almost like there was some bad state left in either Kubernetes or Argo that was removed when I killed the whole argo namespace and re-installed from scratch. I've run through at least 50+ workflows now without re-encountering the issue. Sorry... I wish I could make it happen again. If you want to close this ticket, I can re-open if I run into it again.
